### PR TITLE
ng_sixlowpan: Initial import of IP header compression

### PR DESCRIPTION
**Pull request by authmillenon**
~~Depends on #2614.~~ (merged)

### CODE REVIEW

**Comment by authmillenon** [comment]
Rebased to current dependencies

**Comment by authmillenon** [comment]
Rebased to current master

**Comment by authmillenon** [comment]
Rebased to current #2614.

**Comment by authmillenon** [comment]
Rebased to current master (and removed long out-dated WIP label)

**Comment by OlegHahm** [comment]
Nice. Will review ASAP.

**Comment by authmillenon** [comment]
(please review fragmentation first; I think this is more important right now)

**Comment by OlegHahm** [comment]
Ok

**Comment by authmillenon** [comment]
Rebased to current master

**Comment by OlegHahm** [comment]
Is this ready for review now?

**Comment by authmillenon** [comment]
Yes

**Comment by authmillenon** [comment]
Ever since I opened it ;-)

**Comment by authmillenon** [comment]
Rebased and squashed, since obviously no review took place until now ;-)

**Comment by OlegHahm** [comment]
> Ever since I opened it ;-)

Why was it marked as "WIP" in the beginning then?

**Comment by authmillenon** [comment]
Errrr… I meant when I removed that label ;-)

File: sys/net/network_layer/ng_sixlowpan/iphc/ng_sixlowpan_iphc.c [hunk 1]
```diff
@@ -0,0 +1,695 @@
...
+        return false;
+    }
+
+    return ((ctx->prefix_len == 128) ||
+            ((ctx->prefix_len > 64) &&
+             (memcmp(&(addr->u8[(ctx->prefix_len / 8) + 1]),
+                     &(iid->u8[(ctx->prefix_len / 8) - 7]),
+                     sizeof(network_uint64_t) - ((ctx->prefix_len / 8) - 7)) == 0) &&
+             (addr->u8[(ctx->prefix_len / 8)] & byte_mask[ctx->prefix_len % 8]) ==
+             (iid->u8[(ctx->prefix_len / 8) - 8] & byte_mask[ctx->prefix_len % 8])));
```
Comment (OlegHahm) [inline sys/net/network_layer/ng_sixlowpan/iphc/ng_sixlowpan_iphc.c]: some comments would be useful here

File: sys/net/network_layer/ng_sixlowpan/iphc/ng_sixlowpan_iphc.c [hunk 2]
```diff
@@ -0,0 +1,695 @@
...
+            ng_ipv6_hdr_set_tc(ipv6_hdr, 0);
+            ng_ipv6_hdr_set_fl(ipv6_hdr, 0);
+            break;
+    }
+
+    if (!(iphc_hdr[IPHC1_IDX] & NG_SIXLOWPAN_IPHC1_NH)) {
+        ipv6_hdr->nh = iphc_hdr[payload_offset++];
+    }
+
+    switch (iphc_hdr[IPHC1_IDX] & NG_SIXLOWPAN_IPHC1_NH) {
```
Comment (OlegHahm) [inline sys/net/network_layer/ng_sixlowpan/iphc/ng_sixlowpan_iphc.c]: Shouldn't this check for `NG_SIXLOWPAN_IPHC1_HL` instead?

File: sys/net/network_layer/ng_sixlowpan/iphc/ng_sixlowpan_iphc.c [hunk 3]
```diff
@@ -0,0 +1,695 @@
...
+            ctx = ng_sixlowpan_ctx_lookup_id(sci);
+
+            if (ctx == NULL) {
+                DEBUG("6lo iphc: could not find source context\n");
+                return false;
+            }
+        }
+    }
+
+    switch (iphc_hdr[IPHC2_IDX] & (NG_SIXLOWPAN_IPHC2_SAC | NG_SIXLOWPAN_IPHC2_SAM)) {
```
Comment (OlegHahm) [inline sys/net/network_layer/ng_sixlowpan/iphc/ng_sixlowpan_iphc.c]: Isn't this the same case as in https://github.com/RIOT-OS/RIOT/pull/2806/files#diff-6c12be725eda3c6f60e9b75166d2f4b8R199?

File: sys/net/network_layer/ng_sixlowpan/iphc/ng_sixlowpan_iphc.c [hunk 4]
```diff
@@ -0,0 +1,695 @@
...
+                ipv6_hdr->dst.u8[2] = iphc_hdr[payload_offset++];
+                ipv6_hdr->dst.u8[3] = ctx->prefix_len;
+                ng_ipv6_addr_init_prefix((ng_ipv6_addr_t *)ipv6_hdr->dst.u8 + 4,
+                                         &ctx->prefix, ctx->prefix_len);
+                memcpy(ipv6_hdr->dst.u8 + 12, iphc_hdr + payload_offset + 2, 4);
+
+                payload_offset += 4;
+
+                ctx->prefix_len = orig_ctx_len;
+            } while (0);
```
Comment (OlegHahm) [inline sys/net/network_layer/ng_sixlowpan/iphc/ng_sixlowpan_iphc.c]: I guess it's a sophisticated trick, but I don't quite get it.

File: sys/net/network_layer/ng_sixlowpan/iphc/ng_sixlowpan_iphc.c (hunk 3 above)
Comment (authmillenon) [inline sys/net/network_layer/ng_sixlowpan/iphc/ng_sixlowpan_iphc.c]: L199 is an if-check, preparing for this switch case to reduce code overhead.

File: sys/net/network_layer/ng_sixlowpan/iphc/ng_sixlowpan_iphc.c (hunk 4 above)
Comment (authmillenon) [inline sys/net/network_layer/ng_sixlowpan/iphc/ng_sixlowpan_iphc.c]: It's an ANSI-C compatible block to allocate `orig_ctx_len` locally here.

**Comment by authmillenon** [comment]
Addressed comments

**Comment by authmillenon** [comment]
Squashed

**Comment by OlegHahm** [comment]
untested ACK

**Comment by authmillenon** [comment]
Fixed doc error pointed out by Travis.

**Comment by OlegHahm** [comment]
re-ACK

**Comment by authmillenon** [comment]
Re-Squashed

**Pull request merged**